system/lib/wasmfs/backends/memory_backend.cpp

```diff
@@ -65,15 +65,10 @@ Directory::MaybeEntries MemoryDirectory::getEntries() {
 
 int MemoryDirectory::insertMove(const std::string& name,
                                 std::shared_ptr<File> file) {
-  auto& oldEntries =
-    std::static_pointer_cast<MemoryDirectory>(file->locked().getParent())
-      ->entries;
-  for (auto it = oldEntries.begin(); it != oldEntries.end(); ++it) {
-    if (it->child == file) {
-      oldEntries.erase(it);
-      break;
-    }
-  }
```

Comment:
This part shouldn't need to change because we should never be moving a file between different kinds of backends. What part of this PR made this necessary?

Reply:
I think it's the first `rename()` call in test_fs_icase: https://github.com/emscripten-core/emscripten/blob/00daf403cd09467c6dab841b873710bb878535b2/test/other/test_fs_icase.cpp#L80

We are calling `insertMove()` which goes to the `MemoryDirectory`. But the thing is that the file is actually in `IgnoreCaseDirectory`.
https://github.com/emscripten-core/emscripten/blob/e604df691fc67be040a1862075812c9123998264/system/lib/wasmfs/backends/ignore_case_backend.cpp#L114

Full test error (if I revert changes in MemoryDirectory):

```
test_fs_icase (test_other.other) ... ok
test_fs_icase_wasmfs (test_other.other) ... FAIL

======================================================================
FAIL: test_fs_icase_wasmfs (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/user/src/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/user/src/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 593, in run
    self._callTestMethod(testMethod)
  File "/Users/user/src/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/Users/user/src/emscripten/test/common.py", line 376, in resulting_test
    return func(self, *args)
  File "/Users/user/src/emscripten/test/test_other.py", line 85, in metafunc
    f(self)
  File "/Users/user/src/emscripten/test/test_other.py", line 12537, in test_fs_icase
    self.do_other_test('test_fs_icase.cpp', emcc_args=['-sCASE_INSENSITIVE_FS', '-std=c++20'])
  File "/Users/user/src/emscripten/test/test_other.py", line 161, in do_other_test
    self.do_run_in_out_file_test('other', testname, emcc_args=emcc_args, **kwargs)
  File "/Users/user/src/emscripten/test/common.py", line 1206, in do_run_in_out_file_test
    return self._build_and_run(srcfile, expected, **kwargs)
  File "/Users/user/src/emscripten/test/common.py", line 1250, in _build_and_run
    js_output = self.run_js(js_file, engine, args,
  File "/Users/user/src/emscripten/test/common.py", line 843, in run_js
    self.fail('JS subprocess failed (%s): %s.  Output:\n%s' % (error.cmd, error.returncode, ret))
  File "/Users/user/src/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: JS subprocess failed (/Users/user/src/emsdk/node/14.18.2_64bit/bin/node --stack-trace-limit=50 --unhandled-rejections=throw --trace-uncaught /private/var/folders/c_/szdjrjm12g5_9vh8ybcndhmm0000gp/T/tmp2lk597j2/emscripten_test_other_gye_4dgp/test_fs_icase.js): -10.  Output:
Write 'test' to 'test.txt'
Read 'test' from 'Test.txt'


----------------------------------------------------------------------
Ran 2 tests in 93.604s

FAILED (failures=1)
```

Reply:
Can we layer the virtual backend on top of the memory backend better so that a move in the virtual backend corresponds to a move in the underlying backend rather than a delete and an insert?

Reply:
@tlively How can we do that? We already call `MemoryDirectory::insertMove()`. The problem is that it assumes that the file parent is a `MemoryDirectory` which is not true since the parent is the `IgnoreCaseDirectory`. Actually, the parent of the moved file could be any backend.

Maybe, we could set a parent to be the directory in the underlying backend, then perform move, then change the parent back. But that seem very unreliable.

Reply:
Hi @tlively,
Please, let me know what do you think about my latest comment. Or how else this PR can be improved.
Thanks.

Reply:
Sorry for the long delay here, @mere-human. I'm going to experiment with another way of implementing this backend for a little bit. I'll let you know how it goes.